[BUG] contour algorithm is not working

### Bug Description


```shell
maplibre-gl.KI10pk0I.js:1 Error: AJAXError:  (500): https://titiler-dev.undpgeohub.org/cog/tiles/WebMercatorQuad/6/37/32.png?url=https%3A%2F%2Fundpgeohub.blob.core.windows.net%2Fuserdata%2Fa85516c81c0b78d3e89d3f00099b8b15%2Fdatasets%2FDem_Rwanda_10m_allt_20230921150153.tif%2FDem_Rwanda_10m_allt_20230921150153_band1.tif%3Fc3Y9MjAyMy0wOC0wMyZzcz1iJnNydD1vJnNlPTIwMjQtMTAtMjZUMTAlM0EzMyUzQTA1WiZzcD1yJnNpZz16UG44WFBFME5MeXZ6M0FGYktJdVptZmVxalhsemYzYUlGVm42SmtWd05VJTNE&scale=1&resampling=nearest&return_mask=true&algorithm=contours&algorithm_params={"increment"%3A35%2C"thickness"%3A1%2C"minz"%3A-12000%2C"maxz"%3A8000}&rescale=958.11%2C4123&colormap_name=brbg
at maplibre-gl.KI10pk0I.js:1:7658
at Generator.next (<anonymous>)
at d (maplibre-gl.KI10pk0I.js:1:564)
```

![](https://github.com/UNDP-Data/geohub/assets/2639701/62c74baf-2a02-498e-a9b0-14b569d4c423)

### Bug Reproduction


* add Rwanda dem
* select "contour" for layer type

### Bug Impact
_No response_

### Bug Screenshots
_No response_

### Bug References
_No response_

### Bug System Information


* OS: [e.g. Windows 10]
* Browser: [e.g. Google Chrome, Firefox]



┆Issue is synchronized with this [Wrike task](https://www.wrike.com/open.htm?id=1300324584) by [Unito](https://www.unito.io)

--- a/.changeset/warm-camels-unite.md
+++ b/.changeset/warm-camels-unite.md
@@ -1,0 +1,5 @@
+---
+"geohub": patch
+---
+
+fix: export algorithmId in RasterAlgorithms component to reinitialize layer correctly when algorithm is changed.

--- a/sites/geohub/src/components/maplibre/raster/RasterAlgorithms.svelte
+++ b/sites/geohub/src/components/maplibre/raster/RasterAlgorithms.svelte
@@ -34,6 +34,7 @@
 	export let layerId: string;
 	export let metadata: RasterTileMetadata;
 	export let links: Link[] = [];
+	export let algorithmId = '';
 
 	const isRgbTile = isRgbRaster(metadata.colorinterp);
 
@@ -44,7 +45,6 @@
 			? (metadata.band_metadata.map((meta) => meta[0]) as string[])
 			: [];
 
-	let selectedAlgorithm = '';
 	let algorithms: { [key: string]: RasterAlgorithm };
 
 	let parameters: { [key: string]: number } = {};
@@ -75,7 +75,9 @@
 	const setDefaultParameters = () => {
 		parameters = {};
 
-		if (selectedAlgorithm) {
+		const params = algorithms[algorithmId]?.parameters ?? {};
+
+		if (algorithmId) {
 			let algorithm_params_str =
 				(getValueFromRasterTileUrl($map, layerId, 'algorithm_params') as string) ?? '';
 			if (algorithm_params_str === 'undefined') {
@@ -85,13 +87,13 @@
 				const algorithm_params = JSON.parse(algorithm_params_str);
 				if (Object.keys(algorithm_params).length > 0) {
 					Object.keys(algorithm_params).forEach((key) => {
+						if (!params[key]) return;
 						parameters[key] = algorithm_params[key];
 					});
 				}
 			}
 		}
 
-		const params = algorithms[selectedAlgorithm].parameters;
 		Object.keys(params).forEach((key) => {
 			if (key in parameters) return;
 			parameters[key] = params[key].default;
@@ -103,7 +105,7 @@
 		if (!(layerUrl && layerUrl.length > 0)) return;
 		const layerURL = new URL(layerUrl);
 
-		if (selectedAlgorithm.length === 0) {
+		if (!algorithmId) {
 			layerURL.searchParams.delete('algorithm');
 			layerURL.searchParams.delete('colormap');
 			layerURL.searchParams.delete('colormap_name');
@@ -133,8 +135,8 @@
 		}
 
 		dispatch('change', {
-			id: selectedAlgorithm,
-			algorithm: algorithms[selectedAlgorithm] ?? undefined
+			id: algorithmId,
+			algorithm: algorithms[algorithmId] ?? undefined
 		});
 	};
 
@@ -149,14 +151,12 @@
 		const dumpedParams =
 			Object.keys(parameters).length > 0 ? JSON.stringify(parameters) : undefined;
 
-		let sourceType: 'raster' | 'raster-dem' = ['terrarium', 'terrainrgb'].includes(
-			selectedAlgorithm
-		)
+		let sourceType: 'raster' | 'raster-dem' = ['terrarium', 'terrainrgb'].includes(algorithmId)
 			? 'raster-dem'
 			: 'raster';
 
 		const params = {
-			algorithm: selectedAlgorithm
+			algorithm: algorithmId
 		};
 		if (dumpedParams) {
 			params['algorithm_params'] = dumpedParams;
@@ -179,9 +179,9 @@
 		let oldEncoding = (currentSource as RasterDEMSourceSpecification).encoding ?? '';
 
 		let newEncoding = '';
-		if (selectedAlgorithm === 'terrarium') {
+		if (algorithmId === 'terrarium') {
 			newEncoding = 'terrarium';
-		} else if (selectedAlgorithm === 'terrainrgb') {
+		} else if (algorithmId === 'terrainrgb') {
 			newEncoding = 'mapbox';
 		}
 
@@ -248,7 +248,7 @@
 	};
 
 	const handleParameterValueChanged = () => {
-		if (!selectedAlgorithm) return;
+		if (!algorithmId) return;
 		const layerUrl = getLayerSourceUrl($map, layerId) as string;
 		if (!(layerUrl && layerUrl.length > 0)) return;
 		const layerURL = new URL(layerUrl);
@@ -256,10 +256,10 @@
 	};
 
 	onMount(() => {
-		selectedAlgorithm = (getValueFromRasterTileUrl($map, layerId, 'algorithm') as string) ?? '';
+		algorithmId = (getValueFromRasterTileUrl($map, layerId, 'algorithm') as string) ?? undefined;
 
 		getAlgorithms().then(() => {
-			if (selectedAlgorithm) {
+			if (algorithmId) {
 				setDefaultParameters();
 			}
 		});
@@ -273,8 +273,8 @@
 		</div>
 		<div slot="control">
 			<div class="select is-fullwidth">
-				<select bind:value={selectedAlgorithm} on:change={handleSelectAlgorithm}>
-					<option value="">Use default</option>
+				<select bind:value={algorithmId} on:change={handleSelectAlgorithm}>
+					<option value={undefined}>Use default</option>
 					{#each Object.keys(algorithms) as name}
 						{@const algo = algorithms[name]}
 						{#if algo.inputs.nbands <= availableBands.length}
@@ -286,17 +286,19 @@
 		</div>
 	</FieldControl>
 
-	{#if selectedAlgorithm}
-		{@const params = algorithms[selectedAlgorithm].parameters}
-		{#each Object.keys(params) as key}
-			<RasterAlgorithmParameter
-				bind:id={key}
-				parameter={params[key]}
-				bind:value={parameters[key]}
-				on:change={handleParameterValueChanged}
-				bind:isExpanded={expanded[key]}
-			/>
-		{/each}
+	{#if algorithmId}
+		{@const params = algorithms[algorithmId]?.parameters ?? undefined}
+		{#if params}
+			{#each Object.keys(params) as key}
+				<RasterAlgorithmParameter
+					bind:id={key}
+					parameter={params[key]}
+					bind:value={parameters[key]}
+					on:change={handleParameterValueChanged}
+					bind:isExpanded={expanded[key]}
+				/>
+			{/each}
+		{/if}
 	{/if}
 {:else}
 	<div class="is-flex is-justify-content-center"><Loader size="small" /></div>

--- a/sites/geohub/src/components/maplibre/raster/RasterLegend.svelte
+++ b/sites/geohub/src/components/maplibre/raster/RasterLegend.svelte
@@ -15,6 +15,7 @@
 	export let metadata: RasterTileMetadata;
 	export let tags: Tag[] = [];
 	export let links: Link[] = [];
+	let algorithmId: string = undefined;
 
 	const handleSelectAlgorithm = () => {
 		layerStyle = $map.getStyle().layers.find((l: LayerSpecification) => l.id === layerId);
@@ -50,7 +51,13 @@
 <div class="legend-container">
 	<Accordion title="layer type" bind:isExpanded={expanded['algorithm']}>
 		<div slot="content">
-			<RasterAlgorithms bind:layerId bind:metadata bind:links on:change={handleSelectAlgorithm} />
+			<RasterAlgorithms
+				bind:layerId
+				bind:metadata
+				bind:links
+				on:change={handleSelectAlgorithm}
+				bind:algorithmId
+			/>
 		</div>
 		<div slot="buttons">
 			<Help>Apply an algorithm to visualize the raster dataset as a different layer type.</Help>
@@ -61,7 +68,7 @@
 		<Hillshade bind:layerId />
 	{:else if layerStyle?.type === 'raster'}
 		{#key layerStyle}
-			<RasterLegendEdit bind:layerId bind:metadata bind:tags bind:expanded />
+			<RasterLegendEdit bind:layerId bind:metadata bind:tags bind:expanded bind:algorithmId />
 		{/key}
 	{/if}
 </div>

--- a/sites/geohub/src/components/maplibre/raster/RasterLegendEdit.svelte
+++ b/sites/geohub/src/components/maplibre/raster/RasterLegendEdit.svelte
@@ -45,12 +45,12 @@
 	export let expanded: { [key: string]: boolean } = {
 		color: true
 	};
+	export let algorithmId: string = undefined;
 
 	const isRgbTile = isRgbRaster(metadata.colorinterp);
 	let layerHasUniqueValues = isRgbTile ? false : isUniqueValueRaster(metadata);
 
 	let legendType: LegendType = undefined;
-	let algorithmId: string;
 
 	const unit = tags?.find((t) => t.key === 'unit')?.value;
 
@@ -61,6 +61,7 @@
 	const handleColorMapChanged = () => {
 		if (layerHasUniqueValues) return;
 		if (legendType !== LegendType.LINEAR) return;
+		if (algorithmId) return;
 		// linear colormap
 
 		const currCMAP = getValueFromRasterTileUrl($map, layerId, 'colormap_name') as string;
@@ -95,6 +96,7 @@
 	const handleRescaleChanged = debounce(() => {
 		if (layerHasUniqueValues) return;
 		if (!$rescaleStore) return;
+		if (algorithmId) return;
 		if (legendType !== LegendType.LINEAR) return;
 		const layerStyle = getLayerStyle($map, layerId);
 		const layerUrl = getLayerSourceUrl($map, layerId) as string;
@@ -110,7 +112,9 @@
 	}, 200);
 
 	const decideLegendType = () => {
-		algorithmId = getValueFromRasterTileUrl($map, layerId, 'algorithm') as string;
+		if (!algorithmId) {
+			algorithmId = getValueFromRasterTileUrl($map, layerId, 'algorithm') as string;
+		}
 		if (algorithmId) {
 			legendType = undefined;
 			return;
@@ -136,57 +140,55 @@
 	});
 </script>
 
-{#if !algorithmId}
-	{#if !isRgbTile}
-		<Accordion title="Color" bind:isExpanded={expanded['color']}>
-			<div slot="content">
-				{#if !layerHasUniqueValues}
-					<FieldControl title="Type">
-						<div slot="help">
-							Switch classification type either a simple linear colormap or categorized
-							classification.
-						</div>
-						<div slot="control">
-							<ClassificationSwitch bind:legendType on:change={handleClassificationChanged} />
-						</div>
-					</FieldControl>
-				{/if}
-
-				{#if legendType === LegendType.LINEAR}
-					<div class="field">
-						<div class="control">
-							{#if unit}
-								<span class="unit is-size-6">{unit}</span>
-							{/if}
-							<div class="is-flex">
-								<div style="width: 100%;">
-									<ColorMapPicker
-										bind:colorMapName={$colorMapNameStore}
-										on:colorMapChanged={handleColorMapChanged}
-										isFullWidth={true}
-									/>
-								</div>
-							</div>
-							{#if $rescaleStore?.length > 1}
-								<div class="is-flex">
-									<span class="is-size-6">{$rescaleStore[0].toFixed(2)}</span>
-									<span class="align-right is-size-6">{$rescaleStore[1].toFixed(2)}</span>
-								</div>
-							{/if}
-						</div>
+{#if !algorithmId && !isRgbTile}
+	<Accordion title="Color" bind:isExpanded={expanded['color']}>
+		<div slot="content">
+			{#if !layerHasUniqueValues}
+				<FieldControl title="Type">
+					<div slot="help">
+						Switch classification type either a simple linear colormap or categorized
+						classification.
 					</div>
-				{:else if legendType === LegendType.CATEGORISED}
-					<RasterClassifyLegend bind:layerId bind:metadata />
-				{/if}
-			</div>
-			<div slot="buttons">
-				<Help>Apply a colormap to visualise the raster dataset</Help>
-			</div>
-		</Accordion>
-	{/if}
+					<div slot="control">
+						<ClassificationSwitch bind:legendType on:change={handleClassificationChanged} />
+					</div>
+				</FieldControl>
+			{/if}
+
+			{#if legendType === LegendType.LINEAR}
+				<div class="field">
+					<div class="control">
+						{#if unit}
+							<span class="unit is-size-6">{unit}</span>
+						{/if}
+						<div class="is-flex">
+							<div style="width: 100%;">
+								<ColorMapPicker
+									bind:colorMapName={$colorMapNameStore}
+									on:colorMapChanged={handleColorMapChanged}
+									isFullWidth={true}
+								/>
+							</div>
+						</div>
+						{#if $rescaleStore?.length > 1}
+							<div class="is-flex">
+								<span class="is-size-6">{$rescaleStore[0].toFixed(2)}</span>
+								<span class="align-right is-size-6">{$rescaleStore[1].toFixed(2)}</span>
+							</div>
+						{/if}
+					</div>
+				</div>
+			{:else if legendType === LegendType.CATEGORISED}
+				<RasterClassifyLegend bind:layerId bind:metadata />
+			{/if}
+		</div>
+		<div slot="buttons">
+			<Help>Apply a colormap to visualise the raster dataset</Help>
+		</div>
+	</Accordion>
 {/if}
 
-{#if !layerHasUniqueValues && !isRgbTile}
+{#if !layerHasUniqueValues && !isRgbTile && !algorithmId}
 	<Accordion title="Rescale min/max values" bind:isExpanded={expanded['rescale']}>
 		<div class="pb-2" slot="content">
 			<RasterRescale bind:layerId bind:metadata bind:tags />

